core/src/main/scala/cats/ApplicativeError.scala

```diff
@@ -338,6 +338,130 @@ trait ApplicativeError[F[_], E] extends Applicative[F] {
+   *
+   * Tests an expression, sequencing an exception of type [[java.lang.AssertionError]] if false.
+   *
+   * This method differs from
+   * [[cats.ApplicativeError#assert(assertion:Boolean,message:=>String)* assert]] only in the
+   * intent expressed: `assert` contains a predicate which needs to be proven, while `assume`
+   * contains an axiom for a static checker.
+   *
+   * Assertions can be disabled by using the scalac option `-Xdisable-assertions`, or by
+   * configuring the scalac option `-Xelide-below` with a value greater than
+   * [[scala.annotation.elidable.ASSERTION ASSERTION]].
+   *
+   * When assertions are disabled exceptions are not sequenced, and this method is equivalent to
+   * [[cats.InvariantMonoidal#unit]]
+   *
+   * @param assumption
+   *   the expression to test
+   * @param message
+   *   a String to include in the failure message
+   */
+  final def assume(assumption: Boolean, message: => String)(implicit ev: Throwable <:< E): F[Unit] =
+    catchOnly[AssertionError](Predef.assume(assumption, message))
+
+  /**
+   * Equivalent of a [[scala.Predef#assume(assumption:Boolean)*]] assumption in `F[_]`.
+   *
+   * Tests an expression, sequencing an exception of type [[java.lang.AssertionError]] if false.
+   *
+   * This method differs from [[cats.ApplicativeError#assert(assertion:Boolean)* assert]] only
+   * in the intent expressed: `assert` contains a predicate which needs to be proven, while
+   * `assume` contains an axiom for a static checker.
+   *
+   * Assertions can be disabled by using the scalac option `-Xdisable-assertions`, or by
+   * configuring the scalac option `-Xelide-below` with a value greater than
+   * [[scala.annotation.elidable.ASSERTION ASSERTION]].
+   *
+   * When assertions are disabled exceptions are not sequenced, and this method is equivalent to
+   * [[cats.InvariantMonoidal#unit]]
+   *
+   * @param assumption
+   *   the expression to test
+   */
+  final def assume(assumption: Boolean)(implicit ev: Throwable <:< E): F[Unit] =
+    catchOnly[AssertionError](Predef.assume(assumption))
+
+  /**
+   * Equivalent of a [[scala.Predef#require(requirement:Boolean,message:=>Any)*]] requirement in  `F[_]`.
+   *
+   * Tests an expression, sequencing an exception of type [[java.lang.IllegalArgumentException]]
+   * if false.
+   *
+   * This method is similar to
+   * [[cats.ApplicativeError#assert(assertion:Boolean,message:=>String)* assert]], but blames
+   * the caller of the method for violating the condition.
+   *
+   * Since requirements express a precondition that must be met by the caller, they cannot be
+   * disabled using `-Xdisable-assertions`.
+   *
+   * @param requirement
+   *   the expression to test
+   * @param message
+   *   a String to include in the failure message
+   */
+  final def require(requirement: Boolean, message: => String)(implicit ev: Throwable <:< E): F[Unit] =
+    catchOnly[IllegalArgumentException](Predef.require(requirement, message))
```

Comment:
So I was perusing this PR again and I had another thought. Since this can throw only an `IllegalArgumentException`, what if we weakened the constraint to `IllegalArgumentException <:< E`? The implementation will need some massaging, but it would be more flexible.

I don't have good intuition for typed errors, so not sure whether this makes sense. We can give `AssertionErrors` the same treatment but modeling those seems very strange 😆

Reply:
`catchOnly` requires `Throwable <:< E` doesn't it? And if you have `IllegalArgument <:< E` you don't necessarily have `Throwable <:< E`.

So I don't think that works.

Reply:
Right, that's why I said we'll have to massage the implementation a bit here :) for example, we could inline the implementation of `require` (instead of delegating to `Predef`).